tests/messages/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+#![cfg_attr(not(feature = "std"), feature(alloc_error_handler))]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode};
+#[cfg(not(feature = "std"))]
+use gstd::{prelude::*, *};
+
+#[cfg(feature = "std")]
+#[cfg(test)]
+mod native {
+    include!(concat!(env!("OUT_DIR"), "/wasm_binary.rs"));
+}
+
+#[derive(Encode, Debug, Decode, PartialEq)]
+pub enum Request {
+    SendOnce,
+    SendInf,
+    SendPushAfterCommit,
+    ReplyOnce,
+    ReplyTwice,
+    ReplyPushAfterReply,
+}
+
+#[derive(Encode, Debug, Decode, PartialEq)]
+pub enum Reply {
+    Empty,
+    Error,
+}
+
+#[cfg(not(feature = "std"))]
+mod wasm {
+    extern crate alloc;
+
+    use codec::{Decode, Encode};
+    use gstd::{ext, msg, prelude::*, MessageId, ProgramId};
+
+    use super::{Reply, Request};
+
+    mod gear_sys {
+        extern "C" {
+            pub fn gr_send_init() -> u32;
+            pub fn gr_send_push(handle: u32, data_ptr: *const u8, data_len: u32);
+            pub fn gr_send_commit(
+                handle: u32,
+                message_id_ptr: *mut u8,
+                program: *const u8,
+                gas_limit: u64,
+                value_ptr: *const u8,
+            );
+        }
+    }
+
+    static mut BUFFER: Vec<u8> = Vec::new();
+
+    fn process_request(request: Request) {
+        match request {
+            Request::SendOnce => {
+                msg::send(ProgramId::from(0), "SendOnce", 0, 0);
+            }
+            Request::SendInf => loop {
+                msg::send(ProgramId::from(0), "SendInf", 0, 0);
+            },
+            Request::SendPushAfterCommit => unsafe {
+                let handle = gear_sys::gr_send_init();
+                let mut message_id = [0u8; 32];
+                gear_sys::gr_send_commit(
+                    handle,
+                    message_id.as_mut_ptr(),
+                    ProgramId::from(0).as_slice().as_ptr(),
+                    0,
+                    0u128.to_le_bytes().as_ptr(),
+                );
+                let data = "bytes";
+                gear_sys::gr_send_push(handle, data.as_ptr(), data.len() as u32);
+            },
+            Request::ReplyOnce => {
+                msg::reply("ReplyOnce", 0, 0);
+            }
+            Request::ReplyTwice => {
+                msg::reply("ReplyTwice1", 0, 0);
+                msg::reply("ReplyTwice2", 0, 0);
+            }
+            Request::ReplyPushAfterReply => {
+                msg::reply("ReplyPushAfterReply1", 0, 0);
+                msg::reply_push("ReplyPushAfterReply2");
+            }
+        }
+    }
+
+    #[no_mangle]
+    pub extern "C" fn init() {
+        msg::reply((), 0, 0);
+    }
+
+    #[no_mangle]
+    pub extern "C" fn handle() {
+        msg::load::<Request>()
+            .map(process_request)
+            .unwrap_or_else(|e| {
+                ext::debug(&format!("Error processing request: {:?}", e));
```

Comment:
Prefer `debug!` macro as more readable.

Reply:
I am no longer able to push to this branch, please apply the fix yourself